### PR TITLE
Klaviyo integration

### PR DESCRIPTION
**Pull request by utsabc**
## Description of the change

> Description here

## Type of change
- [ ] Bug fix (non-breaking change that fixes an issue)
- [ ] New feature (non-breaking change that adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)

## Related issues

> Fix [#1]() 

### Development

- [ ] Lint rules pass locally
- [ ] The code changed/added as part of this pull request has been covered with tests
- [ ] All tests related to the changed code pass in development

### Code review 

- [ ]  This pull request has a descriptive title and information useful to a reviewer. There may be a screenshot or screencast attached
- [ ] "Ready for review" label attached to the PR and reviewers mentioned in a comment
- [ ] Changes have been reviewed by at least one other engineer
- [ ] Issue from task tracker has a link to this pull request

---
This change is [Reviewable](https://reviewable.io/reviews/rudderlabs/rudder-sdk-js/212)

### CODE REVIEW

File: integrations/Klaviyo/browser.js [hunk 1]
```diff
@@ -0,0 +1,112 @@
...
+  track(rudderElement) {
+    const { message } = rudderElement;
+    if (message.properties)
+      window._learnq.push(["track", message.event, message.properties]);
+    else window._learnq.push(["track", message.event]);
+  }
+
+  page(rudderElement) {
+    const { message } = rudderElement;
+    if (
```
Comment (sayan-mitra) [inline integrations/Klaviyo/browser.js:95]: change as discussed

**Review by sayan-mitra — CHANGES_REQUESTED** [review]

File: integrations/Klaviyo/browser.js (hunk 1 above)
Comment (utsabc) [reply integrations/Klaviyo/browser.js]: fixed

File: integrations/Klaviyo/browser.js [hunk 2]
```diff
@@ -0,0 +1,112 @@
...
+    } catch (err) {
+      logger.debug(`Error occured at extractCustomFields ${err}`);
+    }
+    window._learnq.push(["identify", payload]);
+  }
+
+  track(rudderElement) {
+    const { message } = rudderElement;
+    if (message.properties)
+      window._learnq.push(["track", message.event, message.properties]);
```
[the remark this answers is not in the document]
Comment (utsabc) [reply integrations/Klaviyo/browser.js]: done

File: integrations/Klaviyo/browser.js [hunk 3]
```diff
@@ -0,0 +1,112 @@
...
+  }
+
+  identify(rudderElement) {
+    const { message } = rudderElement;
+    if (!(message.context && message.context.traits)) {
+      logger.error("user traits not present");
+      return;
+    }
+    let payload = {
+      $id: message.userId,
```
[the remark this answers is not in the document]
Comment (utsabc) [reply integrations/Klaviyo/browser.js]: done

File: integrations/Klaviyo/browser.js [hunk 4]
```diff
@@ -0,0 +1,112 @@
...
+    if (!payload.$email && !payload.$phone_number) {
+      logger.error("user phone or email not present");
+      return;
+    }
+    // Extract other K-V property from traits about user custom properties
+    try {
+      payload = extractCustomFields(
+        message,
+        payload,
+        ["context.traits", "traits"],
```
[the remark this answers is not in the document]
Comment (utsabc) [reply integrations/Klaviyo/browser.js]: done

File: utils/utils.js [hunk 5]
```diff
@@ -491,6 +493,44 @@ function flattenJsonPayload(data) {
...
+// "timezone"])
+// -------------------------------------------
+// The above call will map the fields other than the
+// exlusion list from the given keys to the destination payload
+function extractCustomFields(message, destination, keys, exclusionFields) {
+  keys.map((key) => {
+    const messageContext = get(message, key);
+    if (messageContext) {
+      const values = [];
+      Object.keys(messageContext).map((value) => {
```
[the remark this answers is not in the document]
Comment (utsabc) [reply utils/utils.js]: fixed

File: utils/utils.js [hunk 6]
```diff
@@ -491,6 +493,44 @@ function flattenJsonPayload(data) {
...
+// "image",
+// "timezone"])
+// -------------------------------------------
+// The above call will map the fields other than the
+// exlusion list from the given keys to the destination payload
+function extractCustomFields(message, destination, keys, exclusionFields) {
+  keys.map((key) => {
+    const messageContext = get(message, key);
+    if (messageContext) {
+      const values = [];
```
[the remark this answers is not in the document]
Comment (utsabc) [reply utils/utils.js]: fixed

File: integrations/Klaviyo/browser.js [hunk 7]
```diff
@@ -0,0 +1,122 @@
...
+        : "";
+      const pStr = message.name ? `Page: ${message.name}` : "";
+      const infoString =
+        message.properties.category && message.name
+          ? `${pStr} ${catStr}`
+          : "Page Viewed";
+      if (this.additionalPageInfo && message.properties) {
+        window._learnq.push(["track", `${infoString}`, message.properties]);
+      } else {
+        window._learnq.push(["track", `${pStr} ${catStr}`]);
```
Comment (sayan-mitra) [inline integrations/Klaviyo/browser.js:114]: should be infoString ?

Comment (sayan-mitra) [reply integrations/Klaviyo/browser.js:114]: also, let's do something like ?

```
if (pageCategory && pageName) {
      eventName = `Viewed ${pageCategory} ${pageName} page`;
    }
    
else if (pageName) {
      eventName = `Viewed ${pageName} page`;
    }
else {
      eventName = "Viewed a Page";
    }
```

File: utils/utils.js [hunk 8]
```diff
@@ -491,6 +493,44 @@ function flattenJsonPayload(data) {
 }
 /* ------- End FlattenJson ----------- */
 
+// Extract fileds from message with exclusions
```
Comment (sayan-mitra) [inline utils/utils.js:496]: lets add doc under /** ... */

File: integrations/Klaviyo/browser.js [hunk 9]
```diff
@@ -0,0 +1,122 @@
...
+      logger.error("user phone or email not present");
+      return;
+    }
+    // Extract other K-V property from traits about user custom properties
+    try {
+      payload = extractCustomFields(
+        message,
+        payload,
+        ["context.traits"],
+        [
```
Comment (sayan-mitra) [inline integrations/Klaviyo/browser.js:64]: let's make this a const for the class
and use that in other places?

**Review by sayan-mitra — CHANGES_REQUESTED** [review]

**Comment by sayan-mitra** [comment]
Resolve conflict

File: integrations/Klaviyo/browser.js [hunk 10]
```diff
@@ -0,0 +1,124 @@
...
+        eventName = `Viewed ${message.name} page`;
+      } else {
+        eventName = "Viewed a Page";
+      }
+      if (this.additionalPageInfo && message.properties) {
+        window._learnq.push(["track", `${eventName}`, message.properties]);
+      } else {
+        window._learnq.push(["track", `${eventName}`]);
+      }
+    } else {
```
Comment (sayan-mitra) [inline integrations/Klaviyo/browser.js:118]: is this else block required ?

Comment (sayan-mitra) [reply integrations/Klaviyo/browser.js:118]: as discussed, keeping it is fine

**Review by sayan-mitra — APPROVED** [review]

File: integrations/Klaviyo/browser.js (hunk 10 above)
Comment (utsabc) [reply integrations/Klaviyo/browser.js:118]: resolved as discussed

**Pull request merged**